### ISSUE TITLE
Thoughts on Roles/Permissions modules

### ISSUE THREAD

**Issue by eliascodes** [TwinePlatform/twine-api#91]
In the process of looking at #87, I took another look at the Roles/Permissions modules and had some thoughts.

* [x] Let's make your suggested change of suffixing `Enum` on the end of enumerations
* [x] Inputs and outpus are all in terms of _database IDs_ (implementation detail) instead of _values_ (conceptually more natural). For example it would be good if I could call `Roles.fromUser` and get a value (`'VISITOR'` or `RolesEnum.VISITOR`) instead of a number that describes that value.
* [x] The `ExistsQueryResponse` type should really just be `boolean`
* [x] The `QueryResponse` type doesn't need to be a union type, because `T` in `Promise<T>` is only the type of the resolve value.

Another thing I noticed is that in the Auth strategy definition, the roles/permissions interfaces are being initialised _on each request_. I didn't catch this on review (my bad), but what I intended was for the interfaces each to be initialised once on startup, and then stored to be referenced later (much like the `knex` client is). That was kind of the whole (poorly expressed) point of that pattern; you're binding a variable to a group of functions so that you don't need to pass it in each time.

In general I still kind of like that pattern, however, looking at the code, I can see it actually reduces the possibility for re-use. Example: the `move` method is basically a combination of `userHas`, `remove` and `add`, but it's not possible to call the latter two because they need to be done as a transaction. Passing in the knex client allows us to pass in `Knex.Transaction` instances as well as `Knex` or `Knex.QueryBuilder` instances.

What do you think?

**Comment by eliascodes** [comment]
> if not maybe a higher level auth/types.ts is needed.

I think this is a good idea, since there is obviously some overlap/interrelation.

Will push up my changes shortly for your 👀

**Comment by eliascodes** [comment]
@astroash what do you think about adding a couple of constraints on the `user_account_access_role` table:
* One role per user (`CONSTRAINT UNIQUE (user_account_id, access_role_id)`)
* One org per user (`CONSTRAINT UNIQUE (user_account_id, organisation_id)`)

These are the restrictions I believe we're trying to enforce to begin with to make our lives easier. Making them named constraints means we can add a migration down the line that removes them.

**Comment by astroash** [comment]
I think this depends on a few things:
- how much simpler does it make calling code whilst we are enforcing one role per user
- how long are we intending to keep this restraint, do we imagine removing it when we add the super admin dashboard or further down the line?
- how will removing this restraint in the future affect our codebase? Are we going to be accessing this table in more places than `Roles.oneFromUser`?

**Comment by eliascodes** [comment]
> how much simpler does it make calling code whilst we are enforcing one role per user

Well I think the most significant saving is in terms of data consistency. If the constraints are enforced in the database we can be sure that regardless of what the application code is doing, we won't get additional roles. 

I guess the follow on from that is that we'll need less code doing these checks in the application itself. 

> how long are we intending to keep this restraint, do we imagine removing it when we add the super admin dashboard or further down the line?

I imagine all the design work and UX thinking that needs to go into switching the platform to support multiple orgs and/or roles per user will mean we'll be keeping the constraint at least until after the super-admin stuff is launched. After that it will depend on relative priorities of whatever other new features are nocking around then.

> how will removing this restraint in the future affect our codebase? Are we going to be accessing this table in more places than Roles.oneFromUser?

Probably significantly, but the choice isn't between having the constraint and not having it. The choice is between where the constraint is implemented; application code or in the database.

**Comment by eliascodes** [comment]
I think I marked this as fixed prematurely. @astroash what are your thoughts on my response above?

**Comment by astroash** [comment]
> >how much simpler does it make calling code whilst we are enforcing one role per user
>
> Well I think the most significant saving is in terms of data consistency. If the constraints are enforced in the database we can be sure that regardless of what the application code is doing, we won't get additional roles.

I imagine that the only place we will read from the `user_account_access_role` table is in `Roles.oneFromUser`. In which case we either enforce the one to one relationship in this method or in the database. I'm anti db constraint. I think it will be easier in the long run to adapt `Roles.oneFromUser` and create a new method to get all associated roles when we are at the next stage. 

> > how long are we intending to keep this restraint, do we imagine removing it when we add the super admin dashboard or further down the line?
>
> I imagine all the design work and UX thinking that needs to go into switching the platform to support multiple orgs and/or roles per user will mean we'll be keeping the constraint at least until after the super-admin stuff is launched. After that it will depend on relative priorities of whatever other new features are nocking around then

I think that this feature can be introduced incrementally, rather than all in one go. Allowing users to be assigned to more than one cb is a bigger task than a user having two roles at one cb. eg  allowing a user to be a visitor and volunteer at cb x will not involve design work as they the two roles will be logging into different apps. 

If we add a constraint to the db we will need to make 3 changes to implement the above, add a db migration,  edit `Roles.oneFromUser`, and create a `Roles.fromUser` method. If it is constrained in application code we will only need to do the latter 2. 

**TLDR**
I'm personally in favour of the constraint being implemented in the `Roles.oneFromUser` method rather than in the db.

**Comment by eliascodes** [comment]
> I think that this feature can be introduced incrementally, rather than all in one go.

Yes, but despite this, I can't see it being prioritised before any of the super-admin stuff.

> add a db migration,

This is basically no work. I can write it now:
```sql
ALTER TABLE user_account_access_role
  DROP CONSTRAINT user_account_access_role_one_org_per_user;

ALTER TABLE user_account_access_role
  DROP CONSTRAINT user_account_access_role_one_role_per_user;
```

If that's the only difference between the two, IMO the extra assurance of the DB constraints are well worth it.

**Comment by astroash** [comment]
I don't think the extra assurance is needed if we only access the table from one method. But, I'm not that fussed, and happy with the above suggestion.
